neuralop/models/mwt.py

```diff
@@ -0,0 +1,666 @@
+    
+    def even_odd_reconstruction(self, x):
+        """Dimension-agnostic even-odd reconstruction"""
+        if self.n_dim == 1:
+            B, N, c, ich = x.shape
+            assert ich == 2*self.k
+            x_e = torch.matmul(x, self.rc_e)
+            x_o = torch.matmul(x, self.rc_o)
+            
+            result = torch.zeros(B, N*2, c, self.k, device=x.device)
+            result[..., ::2, :, :] = x_e
+            result[..., 1::2, :, :] = x_o
+            return result
+            
+        elif self.n_dim == 2:
+            B, Nx, Ny, c, ich = x.shape
+            assert ich == 2*self.k**2
+            x_ee = torch.matmul(x, self.rc_ee)
+            x_eo = torch.matmul(x, self.rc_eo)
+            x_oe = torch.matmul(x, self.rc_oe)
+            x_oo = torch.matmul(x, self.rc_oo)
+            
+            result = torch.zeros(B, Nx*2, Ny*2, c, self.k**2, device=x.device)
+            result[:, ::2, ::2, :, :] = x_ee
+            result[:, ::2, 1::2, :, :] = x_eo
+            result[:, 1::2, ::2, :, :] = x_oe
+            result[:, 1::2, 1::2, :, :] = x_oo
+            return result
+            
+        elif self.n_dim == 3:
+            B, Nx, Ny, T, c, ich = x.shape
+            assert ich == 2*self.k**2
+            x_ee = torch.matmul(x, self.rc_ee)
+            x_eo = torch.matmul(x, self.rc_eo)
+            x_oe = torch.matmul(x, self.rc_oe)
+            x_oo = torch.matmul(x, self.rc_oo)
+            
+            result = torch.zeros(B, Nx*2, Ny*2, T, c, self.k**2, device=x.device)
+            result[:, ::2, ::2, :, :, :] = x_ee
+            result[:, ::2, 1::2, :, :, :] = x_eo
+            result[:, 1::2, ::2, :, :, :] = x_oe
+            result[:, 1::2, 1::2, :, :, :] = x_oo
+            return result
+    
+    def forward(self, x):
+        if self.n_dim == 1:
+            B, N, c, ich = x.shape
+            ns = math.floor(np.log2(N))
+        else:
+            B, *spatial_dims, c, ich = x.shape
+            ns = math.floor(np.log2(spatial_dims[0]))
+        
+        Ud = torch.jit.annotate(List[Tensor], [])
+        Us = torch.jit.annotate(List[Tensor], [])
+        
+        for i in range(ns - self.L):
+            d, x = self.wavelet_transform(x)
+            Ud += [self.A(d) + self.B(x)]
+            Us += [self.C(d)]
+        
+        if self.n_dim == 1:
+            x = self.T0(x)
+        else:
+            coarse_size = 2**self.L
+            if self.n_dim == 2:
+                x = self.T0(x.view(B, coarse_size, coarse_size, -1)).view(B, coarse_size, coarse_size, c, ich)
+            elif self.n_dim == 3:
+                T = spatial_dims[2]
+                x = self.T0(x.view(B, coarse_size, coarse_size, T, -1)).view(B, coarse_size, coarse_size, T, c, ich)
+        
+        for i in range(ns - 1 - self.L, -1, -1):
+            x = x + Us[i]
+            x = torch.cat((x, Ud[i]), -1)
+            x = self.even_odd_reconstruction(x)
+        
+        return x
+
+
+class MWT(nn.Module):
```

Comment:
This is great. Let's think of a more descriptive name than MWT, perhaps MultiWaveletNO or MWNO?